### ISSUE TITLE
Stop exposing reviewer names to developers

### ISSUE THREAD

**Issue by bobsilverberg** [mozilla/addons-server#18564]
This is a spin-off from https://github.com/mozilla/addons-server/issues/17318 and is just about removing the names of reviewers from any screens that developers can see, which should be exclusively devhub pages.

**Comment by bobsilverberg** [comment]
As discussed on Matrix, a good approach to identifying when we need to anonymize the name of an `ActivityLog` user would be to use the `action` of the `ActivityLog`. If we can identify an all-encompassing set of actions that would always be *from* a reviewer, then we can anonymize the names for those actions wherever they appear in devhub.

There is an attribute, `reviewer_review_action`, that is assigned to some action classes, which seems like it might be able to serve this purpose. I looked into how this is currently being used in the site.

Here are the action classes that are included, along with their short name:

- APPROVE_VERSION ('Approved')
- PRELIMINARY_VERSION ('Preliminarily approved')
- REJECT_VERSION ('Rejected')
- RETAIN_VERSION ('Retained')
- REQUEST_INFORMATION ('More information requested')
- REQUEST_SUPER_REVIEW ('Super review requested')
- COMMENT_VERSION ('Commented')
- CONFIRM_AUTO_APPROVED ('Auto-Approval confirmed')
- APPROVE_CONTENT ('Content approved')
- REJECT_CONTENT ('Content rejected')
- ADMIN_ALTER_INFO_REQUEST ('Information request altered')
- REQUEST_ADMIN_REVIEW_CODE ('Admin add-on-review requested')
- REQUEST_ADMIN_REVIEW_CONTENT ('Admin content-review requested')
- REQUEST_ADMIN_REVIEW_THEME ('Admin theme-review requested')
- REJECT_CONTENT_DELAYED ('Content reject scheduled')
- REJECT_VERSION_DELAYED ('Rejection scheduled')
- FORCE_DISABLE ('Force disabled')
- FORCE_ENABLE ('Force enabled')

`reviewer_review_action` is used directly in one place, in `reviewers/review.html`, to add a class (`reviewer-review-action`) to an `li` element in “Add-on important changes history”, when the action of the log has this attribute. I couldn’t find the CSS that styles this, but I assume it’s serving some purpose.

Other than that, it is used via the `LOG_REVIEWER_REVIEW_ACTION` constant, which simply collects all of the actions with this attribute into a list. This constant is used in the following four functions, all of which exist inside `ActivityLogManager`:
- `review_log` which returns a queryset of log entries whose actions are in this group, and excludes any entries where the user is `settings.TASK_USER_ID`. This function is used in once place, which is the `reviewlog` view in `reviewers/views.py`. Therefore, LOG_REVIEWER_REVIEW_ACTION controls which log entries appear on this reviewer tools page, and adding an action class to this list (without making any other changes) would cause it to appear in this view.
- `total_ratings` which returns a queryset of values which includes user info and is annotated with `approval_count`, which seems to just be a count of log entries. I’m not sure why it’s called “approval_count” when it seems to include not just approvals but in fact any action that is found in `LOG_REVIEWER_REVIEW_ACTION`. Note that this function does not appear to be used anywhere in the codebase other than in `total_ratings_user_position`, which itself also appears to be unused.
- `monthly_reviews` which is basically the same as `total_ratings`, but just for the most recent month. A fun quirk is the fact that one is called “ratings” while the other is called “reviews’, and I don’t think that either name is accurate. ;-) This is also unused outside of  `monthly_reviews_user_position`, which itself is unused.
- `user_approve_reviews` which returns a queryset of log entries whose actions are in this group, for a specific user.  This is unused outside of  `current_month_user_approve_reviews`, which itself is unused.

If my observations are correct, we can remove `total_ratings`, `total_ratings_user_position`, `monthly_reviews`, `monthly_reviews_user_position`, `user_approve_reviews`, `current_month_user_approve_reviews` and (because it is only used by the preceding functions) `user_position`. This would leave `review_log` as the only place where `LOG_REVIEWER_REVIEW_ACTION` is used.

It looks like none of this is currently used by devhub, which I suppose makes sense as it’s related to reviewers. It’s possible we can reuse `reviewer_review_action` for anonymization in devhub, but I am concerned it might not be all-encompassing enough. Among the actions I think might be missing from the list are:

- APPROVE_VERSION_WAITING ('Approved but waiting')
- REVIEWER_REPLY_VERSION ('Reviewer Reply')
- APPROVAL_NOTES_CHANGED ('Approval notes changed')

I still need to look into where we expose reviewer information in devhub to help decide exactly how to proceed.

**Comment by diox** [comment]
> `reviewer_review_action` is used directly in one place, in `reviewers/review.html`, to add a class (`reviewer-review-action`) to an `li` element in “Add-on important changes history”, when the action of the log has this attribute. I couldn’t find the CSS that styles this, but I assume it’s serving some purpose.

Its usage for `LOG_REVIEWER_REVIEW_ACTION` is what matters here, not the CSS class, but FWIW: https://github.com/mozilla/addons-server/blob/dda718c8f0fa9375e3a673b8468bc2605017e2c2/static/css/zamboni/reviewers.less#L674-L676

**Comment by bobsilverberg** [comment]
> 
> Its usage for `LOG_REVIEWER_REVIEW_ACTION` is what matters here, not the CSS class, but FWIW:
> 
> https://github.com/mozilla/addons-server/blob/dda718c8f0fa9375e3a673b8468bc2605017e2c2/static/css/zamboni/reviewers.less#L674-L676

Thanks @diox. I'm trying to figure out why my code search (I'm doing it with VSCode, my editor) wasn't including that file, but I can't seem to figure that out. I'll keep trying.

Do you agree with my assessment of all of the functions that could be removed? If so, it's probably worth doing so.

**Comment by bobsilverberg** [comment]
> I'm trying to figure out why my code search (I'm doing it with VSCode, my editor) wasn't including that file, but I can't seem to figure that out. 

I guess it was because I was searching for `reviewer_review_action` and not `reviewer-review-action`. 😛

**Comment by bobsilverberg** [comment]
Looking at the usage of reviewer names in devhub, it looks like there are three places where these are exposed, all of which expose them as the user associated with an `ActivityLog` entry:
1. On the Dashboard (at /addons/), there is a panel on the right for "Recent Activity".
2. On the Recent Activity page (/feed/).
3. On the Manage Status and Versions page (/versions/), one can click on "Review History" to see activity.

Interestingly, the first two, which use the same code from [`_get_items`](https://github.com/mozilla/addons-server/blob/master/src/olympia/devhub/views.py#L289) to retrieve `ActivityLog` entries, retrieve all log entries for a list of `addons`, excluding any entries that are in [`LOG_HIDE_DEVELOPER`](https://github.com/mozilla/addons-server/blob/master/src/olympia/constants/activity.py#L808), whereas the third one uses the API implemented via [`VersionReviewNotesViewSet`](https://github.com/mozilla/addons-server/blob/master/src/olympia/activity/views.py#L39) which retrieves all entries for a given `version`, also excluding any entries that are in `LOG_HIDE_DEVELOPER`, but only including entries that are in [`LOG_REVIEW_QUEUE`](https://github.com/mozilla/addons-server/blob/master/src/olympia/constants/activity.py#L800). This combination is in fact implemented as [`LOG_REVIEW_QUEUE_DEVELOPER`](https://github.com/mozilla/addons-server/blob/master/src/olympia/constants/activity.py#L814). It is unclear why these two ways of listing activity for addons/versions retrieve different sets of log actions. For the purposes of this issue it might be worth changing this so they both return the same set? Note that the [`pending_activity_log_count_for_developer`](https://github.com/mozilla/addons-server/blob/master/src/olympia/devhub/templatetags/jinja_helpers.py#L92) also makes use of the `LOG_REVIEW_QUEUE_DEVELOPER` constant.

Because each of the sets of possible `ActivityLog` entries that could be returned is quite large, and many of the actions *could* have been performed by a reviewer, it might be easiest to determine a set of actions that *would* have been performed by a developer, and mark those, and then anonymize the user if the action isn't one of those.

@diox WDYT?

**Comment by diox** [comment]
The 2 sets don't serve the same purposes.
- The dashboard / recent activity page is for all activity. It's meant as a feed or everything that has happened to your add-on, whether it came from you, another developer, a reviewer, etc (excluding of course what we want to record but don't want to show developers). 
- The other is really all about communication with reviewers about a particular version, hence why it focuses on things from `LOG_REVIEW_QUEUE`.

I don't have a strong preference regarding which way to mark the actions.

**Comment by bobsilverberg** [comment]
Based on the above, I think, as suggested, it will be easiest to just identify those actions where we feel we can show the action user's name, and then anonymize everything else. Based on my limited understanding, I think the following are actions where was can show the user's name:
- CREATE_ADDON
- ADD_VERSION
- EDIT_VERSION
- DELETE_VERSION
- CHANGE_LICENSE
- DEVELOPER_REPLY_VERSION
- SOURCE_CODE_UPLOADED
- DEVELOPER_CLEAR_INFO_REQUEST

A couple of questions:
1. Does this seem like an exhaustive list, and would it always be safe to show the name of the user who completed the action in each of those cases?
2. In terms of anonymizing, I was thinking it might be best to just not show anything when anonymizing. I.e., instead of showing "by x", where x is some generic name, we don't show a byline at all.

**Comment by diox** [comment]
1. I suspect there are a lot more actions where we'd show the user name. Glancing over the constants in order, I spot the following missing actions in your list:
- EDIT_PROPERTIES
- EDIT_DESCRIPTION
- EDIT_CATEGORIES
- ADD_USER_WITH_ROLE
- REMOVE_USER_WITH_ROLE
- USER_DISABLE
- USER_ENABLE
- ADD_TAG
- REMOVE_TAG
- CHANGE_USER_WITH_ROLE
- CHANGE_POLICY
- CHANGE_ICON
- APPROVAL_NOTES_CHANGED

I think it's almost always safe to show the name of the user in those cases, but would need @wagnerand's opinion to double-check. One thing I'm wondering about are users with `Addons:Edit` making a change to the add-on in Devhub. It should be extremely rare though, so maybe that's an acceptable edge-case.

2. I question it's a product question... Maybe we'll want "By Mozilla" in some cases ?

**Comment by ioanarusiczki** [comment]
@bobsilverberg 

1. I've checked /developers/feed:
- the name of the reviewer is anonymous "by Add-ons Review Team" at manual approval/rejection
- a reviewer reply is “Reply by reviewer “
- when edits are made from dev hub by someone with edit rights on an add-on the feed will display the name as @diox mentioned it above. 

2. The dev hub /versions page:
- a reviewer info request continues to display the reviewer name, also when the addon is approved manually or rejected the name of the reviewer is displayed in the Review History:

![reviewer name is visible](https://user-images.githubusercontent.com/33448286/151355544-46ac0cd4-ef27-4c68-a4f4-1514900a10e5.png)
![reviewer name](https://user-images.githubusercontent.com/33448286/151355548-e9266977-de92-4169-852b-0e893732d3b4.png)

3. I've tested scenarios for the feed with developer actions, the feed registers them and the name of the developer is displayed.
I think there's an issue when an addon or a new version is submitted,  and a version is deleted/disabled -> the feed would generate 2 notifications with "status changed to Approved" 

![status changed to approved](https://user-images.githubusercontent.com/33448286/151356128-19f8eec1-8bf6-41e9-800b-85491e6c229d.png)

It should be an older issue so I'm going to file it.

Except for that the feed looks good. I'll file a separate issue for /versions/
